test/functional/wallet_avoidreuse.py

```diff
@@ -0,0 +1,149 @@
+    if v < vexp - vspan:
+        raise AssertionError("%s < [%s..%s]" % (str(v), str(vexp - vspan), str(vexp + vspan)))
+    if v > vexp + vspan:
+        raise AssertionError("%s > [%s..%s]" % (str(v), str(vexp - vspan), str(vexp + vspan)))
+
+class AvoidReuseTest(BitcoinTestFramework):
+
+    def set_test_params(self):
+        self.setup_clean_chain = False
+        self.num_nodes = 2
+
+    def reset_balance(self, node, discardaddr):
+        '''
+        Throw away all owned coins by the node so it gets a balance of 0.
+        '''
+        balance = node.getbalance(include_dirty=True)
+        if balance > 0.5:
+            node.sendtoaddress(address=discardaddr, amount=balance, subtractfeefromamount=True, allow_dirty=True)
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        '''
+        Set up initial chain and run tests defined below
+        '''
+
+        self.test_persistence()
+        self.test_immutable()
+
+        self.nodes[0].generate(110)
+        self.sync_all()
+        self.reset_balance(self.nodes[1], self.nodes[0].getnewaddress())
+        self.test_fund_send_fund_send()
+        self.reset_balance(self.nodes[1], self.nodes[0].getnewaddress())
+        self.test_fund_send_fund_senddirty()
+
+    def test_persistence(self):
+        '''
+        Test that wallet files persist the avoid_reuse flag.
+        '''
+        # Configure node 1 to use avoid_reuse
+        self.nodes[1].setwalletflag('avoid_reuse')
+
+        # Flags should be node1.avoid_reuse=false, node2.avoid_reuse=true
+        assert_equal(self.nodes[0].getwalletinfo()["avoid_reuse"], False)
+        assert_equal(self.nodes[1].getwalletinfo()["avoid_reuse"], True)
+
+        # Stop and restart node 1
+        self.stop_node(1)
+        self.start_node(1)
+        connect_nodes_bi(self.nodes, 0, 1)
+
+        # Flags should still be node1.avoid_reuse=false, node2.avoid_reuse=true
+        assert_equal(self.nodes[0].getwalletinfo()["avoid_reuse"], False)
+        assert_equal(self.nodes[1].getwalletinfo()["avoid_reuse"], True)
+
+    def test_immutable(self):
+        '''
+        Test immutable wallet flags
+        '''
+        # Attempt to set the disable_private_keys flag; this should not work
+        assert_raises_rpc_error(-8, "Wallet flag is immutable", self.nodes[1].setwalletflag, 'disable_private_keys')
+
+        tempwallet = ".wallet_avoidreuse.py_test_immutable_wallet.dat"
+
+        # Create a wallet with disable_private_keys set; this should work
+        self.nodes[1].createwallet(tempwallet, True)
+        w = self.nodes[1].get_wallet_rpc(tempwallet)
+
+        # Attempt to unset the disable_private_keys flag; this should not work
+        assert_raises_rpc_error(-8, "Wallet flag is immutable", w.setwalletflag, 'disable_private_keys', False)
+
+        # Unload temp wallet
+        self.nodes[1].unloadwallet(tempwallet)
+
+    def test_fund_send_fund_send(self):
```

Comment:
The way I see it is, the two tests above it will catch weird issues that could cause the test to fail in random places, so those tests are run first. The intention is to order tests chronologically.